gax-java/gax/src/main/java/com/google/api/gax/rpc/ApiCallContext.java

```diff
@@ -266,4 +272,21 @@ public static <T> Key<T> create(String name) {
       return new Key<>(name);
     }
   }
+
+  /**
+   * Validate the Universe Domain to ensure that the user configured Universe Domain and the
+   * Credentials' Universe Domain match. An exception will be raised if there are any issues when
+   * trying to validate (i.e. unable to access the universe domain).
+   *
+   * <p>For both transports, a PermissionDenied exception will be thrown if the universe domain is
+   * invalid
+   *
+   * <p>For gRPC, if the client library is unable to retrieve the Universe Domain, the client
+   * library will attempt to retry with the defined retry bounds. If the retry bounds have been
+   * exceeded, the client library will return an Unavailable exception back.
+   *
+   * <p>For HttpJson, if the client library is unable to retrieve the Universe Domain, the client
+   * library will fail and return an exception.
+   */
+  void validateUniverseDomain();
```

Comment:
Possible future enhancement: Both transports should have the same implementation for validating the universe domain. Migrate the transport specific validation code here and provide a default implementation in ApiCallContext (transport-agnostic).

Reply:
Agree with the enhancement. I guess the reason we don't do it now is we don't have the getters for `credentials` in `ApiCallContext`?

Reply:
The reason it's not done now is because gRPC and HttpJson have minor behavior differences regarding how to retry the Auth Exception (marking as Unavailable and retrying vs throwing an exception). In a future enhancement, gRPC and HttpJson should have the same behavior w.r.t. dealing with access token + universe domain failures.